Tooling setup for Clair V4 testing

vishnuchalla opened the pull request:
### Description
Tooling setup to test Clair V4
### List of Apis
```
POST /indexer/api/v1/index_report
GET /indexer/api/v1/index_report/{manifest_hash}
DELETE /indexer/api/v1/index_report/{manifest_hash}
GET /indexer/api/v1/index_state
GET /matcher/api/v1/vulnerability_report/{manifest_hash}
```
HTTP workload generator tool:  [vegeta](https://github.com/tsenart/vegeta)
Results are published to local `./logs` directory and indexed in elastic search instance as well.
### Testing
Tested deploying the below yaml file on openshift cluster: https://gist.github.com/vishnuchalla/f47ba92ad57a30aa86947f6e9fcfeb40
Results:
Complete log: https://gist.github.com/vishnuchalla/dacbb6dde7e8ceecb427563972aec4c6
Elastic search docs: https://gist.github.com/vishnuchalla/5486637b34b0d6b53cab2715e6ee4767

Commit:
testing with vegeta

Commit:
making blocks for code ready for clair

Commit:
added indexing and logging info and stderr

Commit:
finalizing to use zlog and urfave CLI

Commit:
package things for modularity

Commit:
abstracting out logic and enforcing latet go

Commit:
adding make file and generating binary outof it

Commit:
correcting repo typo in makefile

Commit:
fixing lint bugs and finalizing MakeFile

Commit:
corrected version in make file and added all apis

Commit:
separating out manifests

Commit:
working on core logic to run load

Commit:
adding a load script as asset for loading images

Commit:
initial draft of tests

Commit:
handling exception for manifest multiprocessing code

Commit:
splitting out content into files

Commit:
refactoring attacker and indexing

Commit:
fixing a typo in the attacker module

Commit:
testing setup ready to test. need to containerise

Commit:
containerised the test setup

Commit:
updating assets and README.md

hdonnay commented on go.mod:
```diff
@@ -1,16 +1,33 @@
-module github.com/crozzy/clair-load-test
+module github.com/vishnuchalla/clair-load-test
```
Why is the module renamed into your personal fork?

vishnuchalla replied:
This is just for testing purpose as some of the modules are missing from crozzy's code. Once merged we can change it.

> This all just leaves me with a lot of questions:
> 
> * Why is elasticsearch a dependency?
We are trying to store our end results to ES. The advantage of this is we can easily create grafana dashboards and compare between two run ids for our analysis. (will be useful for automation)
> * Why is the Makefile wholly rewritten?
The previous Makefile has code just to build an executable binary. But the new one has functionalities like `go linting`, `go fmt`, `image build`, `image push` etc. Which will help us have an executable in the `bin` directory as well as it pushed to an image registry. This will help us containerize our code and run on openshift so that we don't have to worry about dependency changes often. (will again be useful for automation standpoint). And also all the stuff is done through a single command `make`.
> * Why are so many files shuffled around into their own packages?
The idea is to make code more modular and break it down into individual components rather than having everything in a single large file. This will make our job easier with `debugging`, `adding new functionality` and `maintaining` the code base.
> * Why is the tool built outside the Containerfile and copied in?
>  So, we are trying to provide flexibilty by both CLI version (with executable binary) and containerized version(which will run on openshift) to the end user. Since we already have executable built, we are simply copying it into the container and using the same to trigger tests.
> At the very least, this needs `go fmt`, the `Containerfile` cleaned up, the Makefile changes undone (IMO -- crozzy make care less), and elasticsearch made strictly optional.
> * `go fmt` - Done
> * `Containerfile cleanup` - Done
> * `Makefile cleanup` - Done
And elastic search is required as stated above. Feel free to let me know if you have any further suggestions/concerns. 
> I guess I'll wait for @crozzy to chime in because this is his brainchild, but this is unmergeable as-is, IMO.

Hi @hdonnay - Thanks for a quick review. Here are the inline answers for your questions.

hdonnay replied:
- Yeah, we shouldn't make elasticsearch a hard dependency. It should just get used if the information is provided.
- The makefile doesn't add anything -- all those things are single commands and they're all `PHONY` targets.
- Shuffling the code around doesn't actually help for those things, since they're used in exactly one place.

hdonnay replied:
You should fix the name before we merge this, then we don't have to change it again.

vishnuchalla replied:
> * Yeah, we shouldn't make elasticsearch a hard dependency. It should just get used if the information is provided.
Acknowledged. Changed accordingly. Elastic Search is optional now.
> * The makefile doesn't add anything -- all those things are single commands and they're all `PHONY` targets.
Cleanedup the make file and kept only the targets that I felt necessary. Please let me know if you have any questions/concerns with the updated file. Thanks!
> * Shuffling the code around doesn't actually help for those things, since they're used in exactly one place.
Acknowledged. The idea was to modularize it. Followed Crozzy's suggestions below to make those changes. Feel free to let me know if you have any further concerns. Thanks.

Here are my inline answers to the above questions.

hdonnay commented on Containerfile:
```diff
@@ -0,0 +1,55 @@
+WORKDIR /tmp
+ARG DEBIAN_FRONTEND=noninteractive
+
+# ALPINE
```
???

vishnuchalla replied:
Thanks for pointing out. I have been trying out multiple things and at the end forgot to cleanup. I have updated it now. Please let me know if you have any questions in specific on any line. In summary it installs

* vegeta - Used for http workload generating.
* snafu - Used to process vegeta results and publish them to ES.
* clairctl - Used to fetch manifest of containers and use them during testing.

And at the end it cleans up some installation cache and triggers the command.

hdonnay reviewed (changes requested):
This all just leaves me with a lot of questions:

- Why is elasticsearch a dependency?
- Why is the Makefile wholly rewritten?
- Why are so many files shuffled around into their own packages?
- Why is the tool built outside the Containerfile and copied in?

At the very least, this needs `go fmt`, the `Containerfile` cleaned up, the Makefile changes undone (IMO -- crozzy make care less), and elasticsearch made strictly optional.

I guess I'll wait for @crozzy to chime in because this is his brainchild, but this is unmergeable as-is, IMO.

Commit:
resolving comments based on initial review

crozzy commented on pkg/utils/types.go:
```diff
@@ -0,0 +1,16 @@
+	Host           string   `json:"host"`
+	HitSize        int      `json:"hitsize"`
+	IndexDelete    bool     `json:"delete"`
+	Psk            string   `json:"-"`
```
`PSK`

vishnuchalla replied:
Changed as suggested. Thanks!

crozzy commented on pkg/utils/types.go:
```diff
@@ -0,0 +1,16 @@
+	HitSize        int      `json:"hitsize"`
+	IndexDelete    bool     `json:"delete"`
+	Psk            string   `json:"-"`
+	Uuid           string   `json:"uuid"`
```
`UUID`

vishnuchalla replied:
Changed as suggested. Thanks!

crozzy commented on pkg/utils/utils.go:
We shouldn't have a `utils` package, `GetRequestCommons` should live where it's called (in the report code), IMO `GetContainersList` shouldn't be a function, it's more code and harder to follow at a glace.

vishnuchalla replied:
Acknowledged. My idea was to break everything into small chunks and make it extensible with new features. Since we have limited functionality as of now, moving these two to their respectively caller files as private methods and removing utils package. Thanks

crozzy commented on pkg/attacker/attacker.go:
These shouldn't be in `pkg`, it's not library code and wouldn't ever be imported by another project, if you really want to modularize then root level packages are fine.

vishnuchalla replied:
Acknowledged. Changed accordingly. Thank you

crozzy commented on cmd/clair-load-test/clair-load-test.go:
```diff
@@ -0,0 +1,80 @@
+	"github.com/quay/zlog"
+	"github.com/rs/zerolog"
+	"github.com/urfave/cli/v2"
+	"github.com/vishnuchalla/clair-load-test/pkg/reports"
```
Obviously these imports need to be changed before merging, my bad if they were `crozzy` before, must have missed that when we migrated namespace

vishnuchalla replied:
Acknowledged. I will change it at the end just before merged into main. Because changing it before will get us errors due to this new code structure not being present in the previous code.

crozzy replied:
I believe it doesn't work now because you also changed the go.mod "module" directive from which the import path is dictated. Changing everything to github.com/quay/clair-load-test should be good

crozzy commented on pkg/reports/reports.go:
```diff
@@ -0,0 +1,162 @@
+	var requests []map[string]interface{}
+	var testName string
+	requests, testName = CreateIndexReport(ctx, manifests, conf.Host, jwt_token)
+	attacker.RunVegeta(requests, testName, conf)
```
Why not return errors here instead of panicking? Also, the load test should be able to be run without an elasticsearch connection.

vishnuchalla replied:
Made changes to return errors. And also made elastic search optional, It will only execute when a ESURL is specified. In case of faulty ES details nothing breaks, we will have output to the console as well as to our log dir in json format.

crozzy commented on pkg/reports/reports.go:
```diff
@@ -0,0 +1,162 @@
+	zlog.Info(ctx).Str("Uuid", conf.Uuid).Msg("Run details")
+	var requests []map[string]interface{}
+	var testName string
+	requests, testName = CreateIndexReport(ctx, manifests, conf.Host, jwt_token)
```
returning testName here seems useless and an antipattern

vishnuchalla replied:
Thanks for pointing it out. Removed it.

crozzy commented on pkg/token/token.go:
```diff
@@ -54,10 +63,20 @@ func createToken(key string) (string, error) {
+}
+
+// Method to fetch signing key.
+func getSigningKey(key string) ([]byte, error) {
```
Useless function

vishnuchalla replied:
Removed! Thanks

crozzy commented on pkg/token/token.go:
```diff
@@ -54,10 +63,20 @@ func createToken(key string) (string, error) {
 }
+
+// Method to get signing algorithm for token.
+func getSigningAlgorithm() jose.SignatureAlgorithm {
```
Useless function

vishnuchalla replied:
Removed. Thanks!

crozzy commented on pkg/attacker/attacker.go:
```diff
@@ -0,0 +1,122 @@
+	"strings"
+)
+
+// Method to generate requests string to feed it as input to vegeta.
```
None of these functions are methods (this applies to most of the code), this documentation is appreciated but we should start with the proper [format](https://go.dev/blog/godoc)

vishnuchalla replied:
Changed as suggested. Thanks!

crozzy commented on pkg/attacker/attacker.go:
```diff
@@ -0,0 +1,122 @@
+	cmd.Stdout = os.Stdout
+	err = cmd.Run()
+	if err != nil {
+		panic(err)
```
Strange to panic here

vishnuchalla replied:
Modified. Thanks

crozzy commented on pkg/manifests/manifests.go:
```diff
@@ -0,0 +1,59 @@
+			defer mu.Unlock()
+			listOfManifests[i] = manifest
+			listOfManifestHashes[i] = blob.ManifestHash
+			results <- result{index: i}
```
Your results chan is essentially an errChan as that's all you check so there's no point in this line, all of this concurrency would be cleaner with and `errGroup`.

vishnuchalla replied:
Yes thought in that direction as well. But the problem with errgroup is, it only returns the first error with exit code 1 and skips remainings: https://github.com/golang/go/issues/23595. Where in this approach gets all the errors and gracefully skips to others if fetching a tag fails, which is required in somecases. Thanks.

log with errgroup (where we have 5 invalid manifests to fetch)
```
2023-04-13T07:13:43-04:00 DBG Error generating manifests: could not generate manifest for container quay.io/vchalla/clair-load-test:mysql_8.0.2_tag_11: exit status 1
```
log with this new approach  (where we have 5 invalid manifests to fetch)
```
2023-04-13T07:32:28-04:00 DBG Error generating manifest for container. Message: could not generate manifest: exit status 1 container=quay.io/vchalla/clair-load-test:mysql_8.0.2_tag_13
2023-04-13T07:32:28-04:00 DBG Error generating manifest for container. Message: could not generate manifest: exit status 1 container=quay.io/vchalla/clair-load-test:mysql_8.0.2_tag_4
2023-04-13T07:32:28-04:00 DBG Error generating manifest for container. Message: could not generate manifest: exit status 1 container=quay.io/vchalla/clair-load-test:mysql_8.0.2_tag_17
2023-04-13T07:32:28-04:00 DBG Error generating manifest for container. Message: could not generate manifest: exit status 1 container=quay.io/vchalla/clair-load-test:mysql_8.0.2_tag_15
2023-04-13T07:32:28-04:00 DBG Error generating manifest for container. Message: could not generate manifest: exit status 1 container=quay.io/vchalla/clair-load-test:mysql_8.0.2_tag_24
```

crozzy commented on pkg/manifests/manifests.go:
```diff
@@ -0,0 +1,59 @@
+	go func() {
+		for res := range results {
+			if res.err != nil {
+				zlog.Debug(ctx).Str("container", res.container).Msg(fmt.Sprintf("Error generating manifest for container. Message: %v", res.err))
```
Do we want to continue after this error?

vishnuchalla replied:
Yes. The intent is to skip the failed ones and move forward with the other containers to load test. Thanks!

crozzy replied:
I'm not sure about this, feels like for a loadtest we want to try and have reasonable parity of input between runs. i.e if we fail to get the manifest for a larger image it's going to skew the results for average latency

vishnuchalla replied:
No, there is no chance for that scenario to occur. What we are doing in this load test is, fetch manifests of all the containers at once and then trigger workload with the available list of manifests on the API endpoints. Even if an image size is larger and fetching manifest takes time, it shouldn't affect our load test on the API endpoints as it is all-together a separate step conducted as a pre-run. And also in this new approach I have provided a flexibility to create image manifests all of same size with a load script [here](https://github.com/vishnuchalla/clair-load-test/blob/vchalla/assets/image_load.sh). And once the load script is created using a specific image with a prefix we can have multiple images of same size ready for our run phase like in [here](https://quay.io/repository/vchalla/clair-load-test?tab=tags).

crozzy replied:
So you're saying this scenario cannot occur:
* Run 1, gets to this step tries to get 100 manifests, fails on 50 of them so IIUC the loadtest will begin with these 50 manifests
* Run 2, gets to this step and tries to get 100 manifests, fails on 50 (a different 50 that incidentally have a smaller layers)
* When comparing the results will run 2 not seems slower than run 1 as the inputs are different even though they we specified as the same?

IMO we should error here, it's a rare situation that getting a manifest definition fails and it's worth exploring if it does.

vishnuchalla replied:
Yes, with these new code changes in place, this should be fixed by now. Thanks!

crozzy commented on pkg/reports/apis.go:
I general I like this idea of calling the API directly

vishnuchalla replied:
Thank you!

crozzy commented on pkg/reports/apis.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+// Method to test POST operation in index report.
+func CreateIndexReport(ctx context.Context, manifests [][]byte, host, token string) ([]map[string]interface{}, string) {
```
`CreateIndexReportRequests` no? (same for the others)

vishnuchalla replied:
Changed accordingly. Thanks!

crozzy commented on pkg/reports/apis.go:
```diff
@@ -0,0 +1,85 @@
+
+// Method to test GET operation in index report.
+func GetIndexReport(ctx context.Context, manifestHashes []string, host, token string) ([]map[string]interface{}, string) {
+	zlog.Debug(ctx).Str("requests", strconv.Itoa(len(manifestHashes))).Msg("preparing requests for GET operation in index_report")
```
Don't use the `Str()` method here

vishnuchalla replied:
Done. Thanks

crozzy commented on pkg/attacker/attacker.go:
```diff
@@ -0,0 +1,122 @@
+		req["url"] = reqDict["url"]
+		req["method"] = reqDict["method"]
+		// Encode the body as base64-encoded JSON if the request method is not GET
+		if reqDict["body"] != nil && reqDict["method"] != "GET" {
```
DELETE (also it's good practice to use the `http` [constants](https://pkg.go.dev/net/http#pkg-constants))

vishnuchalla replied:
Acknowledged. Thanks!

hdonnay commented on Containerfile:
```diff
@@ -1,54 +1,38 @@
-                       postgresql-client libffi-dev libxml2 \
-                       libxml2-dev libxslt-dev libjpeg-dev \
-                       zlib1g-dev musl-dev
+RUN apt-get install -y podman wget git dumb-init python3.6 python3-distutils python3-pip python3-apt
```
Don't `update` and `upgrade` -- you should rely on the image you're building on being up to date.

All the other `apt-get` commands need to happen in one step -- running `apt-get clean` only makes the container image larger.

vishnuchalla replied:
I am sorry, we will have to run update in order for all the dependent python packages to get installed. And also it is stated as one of the best practices: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run here. Thanks!

hdonnay commented on Containerfile:
```diff
@@ -1,54 +1,38 @@
+# Install vegeta for HTTP benchmarking
 RUN wget https://github.com/tsenart/vegeta/releases/download/v12.8.3/vegeta-12.8.3-linux-amd64.tar.gz
 RUN tar -xzf vegeta-12.8.3-linux-amd64.tar.gz
 RUN mv vegeta /usr/local/bin/vegeta
```
All these need to happen in one step, and the tarball needs to be deleted.

vishnuchalla replied:
Changed as suggested. Thanks!

crozzy commented on pkg/attacker/attacker.go:
```diff
@@ -0,0 +1,122 @@
+	log_directory := "./logs"
+	if _, err := os.Stat(log_directory); os.IsNotExist(err) {
+		if err := os.MkdirAll(log_directory, 0755); err != nil {
+			// Handle the error here
```
I agree ;)

hdonnay commented on Containerfile:
```diff
@@ -1,54 +1,38 @@
+# Install clairctl to get clair manifests
 RUN wget https://github.com/quay/clair/releases/download/v4.6.0/clairctl-linux-amd64
 RUN chmod 777 clairctl-linux-amd64
 RUN mv clairctl-linux-amd64 /usr/local/bin/clairctl
```
All these need to happen in one step

vishnuchalla replied:
Changed as suggested. Thanks!

crozzy commented on pkg/attacker/attacker.go:
```diff
@@ -0,0 +1,122 @@
+	if err != nil {
+		panic(err)
+	}
+	fmt.Printf("Results for test %s written to file: %s\n", testName, result_filename)
```
log? (and other places in this file)

vishnuchalla replied:
Modified. Thanks!

hdonnay commented on Containerfile:
```diff
@@ -1,54 +1,38 @@
 RUN cd benchmark-wrapper && \
     cp -R . /opt/snafu/
 RUN pip3 install --upgrade pip
 RUN pip3 install -e /opt/snafu/
```
All this needs to happen in one step. Why is the repo being copied? This would be better written as downloading a snapshot from gihub and untaring it into place.

vishnuchalla replied:
Done as suggested. Thanks!

hdonnay commented on Containerfile:
```diff
@@ -1,54 +1,38 @@
+# Cleanup the installation remainings
 RUN apt-get clean autoclean
 RUN apt-get autoremove --yes
 RUN rm -rf /var/lib/{apt,dpkg,cache,log}/
```
this `rm` isn't helpful

vishnuchalla replied:
The above 3 lines are useful for the cleanup and they will reduce some amount of container size.

* The first command `RUN apt-get clean` autoclean removes packages that were automatically installed to satisfy dependencies and are no longer needed. This frees up disk space in the container.
* The second command `RUN apt-get autoremove --yes` removes packages that were installed as dependencies for other packages, but are no longer needed by any package. This also frees up disk space.
* The third command `RUN rm -rf /var/lib/{apt,dpkg,cache,log}/` removes cached package files, temporary files, and logs from the `/var/lib/apt, /var/lib/dpkg, and /var/log` directories. This helps reduce the size of the container and keeps it clean.

Please feel free to share your concerns on this. Thank you!

jtaleric replied:
we usually attempt to reduce the image size for most of our container builds -- if there is a better way to do this, let us know!

hdonnay replied:
You need to remove files in the same step they're created, otherwise they just make the container larger.

vishnuchalla replied:
Acknowledged. Moved the cleanup commands along with `apt-get update` and `apt-get install` as suggested. Thanks!

hdonnay reviewed (changes requested):
I think this PR should just be abandoned and done in a much smaller, piecewise fashion.

crozzy reviewed (commented):
A little context here:
In order for a service to be onboarded by app-sre there needs to have been load-testing, the results of which are added to app-interface in documentation. Another requirement is that the load-testing can re-run by someone unfamiliar with the intricacies of the project (and those steps are described in an SOP). Given those requirements I built this small application with very limited dependencies where the reproduction steps are clearly documented and tested and the output is available w/o digging. These requirement must stay in place. Essentially the project is a small wrapper around clairctl.

vishnuchalla commented:
> I think this PR should just be abandoned and done in a much smaller, piecewise fashion.

I am afraid, by doing this it might break the existing tool functionality. And also If broken into small PRs what would be the plan to get parts of it merged? I prefer to have this PR merged and take changes on top it in a small piece fashion. But open to other suggestions as well. Thank you

Commit:
resolving comments on PR-1

Commit:
cleaning up make file and changing package name

vishnuchalla commented:
> A little context here: In order for a service to be onboarded by app-sre there needs to have been load-testing, the results of which are added to app-interface in documentation. Another requirement is that the load-testing can re-run by someone unfamiliar with the intricacies of the project (and those steps are described in an SOP). Given those requirements I built this small application with very limited dependencies where the reproduction steps are clearly documented and tested and the output is available w/o digging. These requirement must stay in place. Essentially the project is a small wrapper around clairctl.

Yes. Understood. I have made elastic search optional. With the new changes, someone unfamiliar with the tool can simply execute the tool and get the results in stdout and logs directory as well. And also if elastic search connection details are specified, the same results will be indexed too. Thanks!

vishnuchalla committed:
deleting snafu snapshot file after installation

vishnuchalla committed:
performing cleanup along with apt-get update & install

crozzy commented on attacker/attacker.go:
Looks like vegeta is a go project with an exposed API https://github.com/tsenart/vegeta#usage-library Why not use the library?

vishnuchalla replied:
Yes, I have looked at this option as well. Both the `CLI` and `library` have different version releases and are not inline with each other. The latest `library` release works but doesn't have some import features like `lazy` attack option (used to simulate a real time load test) and has some GO limits regarding number of file descriptors open and threads in parallel (when it comes to concurrency, i.e rate). And also test `duration` is mandatory in `library` version which we don't wanted here to simulate a real time work load. So, we have decided to use the `CLI` version which has all the required options and has been very stable since years for all our test runs.

crozzy replied:
https://github.com/tsenart/vegeta/blob/master/attack.go#L175 It appears that the cli is calling the library so there should be a way to achieve the same behavior (i.e. https://github.com/tsenart/vegeta/blob/master/attack.go#L175 `When the duration is zero the attack runs until Stop is called`, it also looks like lazy is the default when using the lib https://github.com/tsenart/vegeta/blob/master/attack.go#L156)

vishnuchalla replied:
Changed as suggested implemented vegeta GO library for achieving same functionality.

crozzy commented on pkg/manifests/manifests.go:
This code is still inside a `pkg` dir, but it's not library code that could be externally exportable, as far as I can see, no code in this PR comes under that definition

vishnuchalla replied:
Acknowledged. My only concern was to modularize the code instead of having the entire code in one large file. Would moving all those files to root level and removing `pkg` be okay?

crozzy replied:
There is no constrain to put all the code in one large file. A project should grow and modularize naturally as the functionality grows (as Hank alluded to), doing it preemptively only serves to make it less readable. Your changes make it difficult to navigate for someone coming to the project. If you're creating a `cmd/clair-load-test` dir you should put the `cli.Command` definitions in there (see: https://github.com/quay/clair/tree/main/cmd/clairctl). The vegeta stuff could probably exist in it's own root level package as it is now.

vishnuchalla replied:
Okay, Understood. Maybe I was thinking too far. Will make changes accordingly and update the PR.

crozzy commented on pkg/commons/types.go:
No commons package, if it's holding config call it config (or as hank mentioned, don't modularize everything preemptively), also this looks specific to the `report` command

vishnuchalla replied:
Removed commons package and changed as suggested.

crozzy commented on attacker/attacker.go:
```diff
@@ -0,0 +1,143 @@
+
+// writeVegetaResults writes all the vegeta output to the specified log directory.
+// It returns the filename where results are logged and errors if any.
+func writeVegetaResults(ctx context.Context, UUID, testName string, vegetaOutput []byte) (string, error) {
```
nit: if it's a UUID pass it as that type

vishnuchalla replied:
Explanation here: https://github.com/quay/clair-load-test/pull/1#discussion_r1174597988. But open to other thoughts too.

crozzy replied:
Thanks, in that case it should not be called `UUID` as a UUID has a specific format,

crozzy commented on pkg/reports/reports.go:
```diff
@@ -0,0 +1,191 @@
+		&cli.StringFlag{
+			Name:    "uuid",
+			Usage:   "--uuid f519d9b2-aa62-44ab-9ce8-4156b712f6d2",
+			Value:   uuid.New().String(),
```
This default doesn't make sense when the help is rendered: 
```
(default: "14484a83-abba-483c-9b66-3b5ce93b4088")
```
Also, what is the point in making this a flag? For me the UUID of the test should be created by the application and surfaced to the user in the output

jtaleric replied:
In some cases the user might want to override the generated uuid, to help do a/b comparisons. ie `joes-fips-run-rhel9-aws`

hdonnay replied:
That's not a uuid -- is this just some ID for the run?

vishnuchalla replied:
Yes. If a user specifies it, it is some ID to run. But if not specified, it will be randomly generated unique ID. We are providing this option, so that the user can easily identify the results of their run using this ID either in local or ES.

crozzy commented on attacker/attacker.go:
```diff
@@ -0,0 +1,143 @@
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/quay/clair-load-test/pkg/commons"
```
nit: imports not properly formatted.

vishnuchalla replied:
Corrected it. Thank you

crozzy commented on pkg/reports/reports.go:
```diff
@@ -0,0 +1,191 @@
+	"encoding/json"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/quay/clair-load-test/attacker"
```
nit: imports not correctly formatted

vishnuchalla replied:
Corrected it. Thank you

crozzy commented on pkg/manifests/manifests.go:
```diff
@@ -0,0 +1,61 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/quay/zlog"
```
nit: imports not correctly formatted

vishnuchalla replied:
Corrected it. Thanks!

crozzy reviewed (changes requested):
Some more comments

vishnuchalla committed:
fixing PR comments and enabling batch processing for manifest fetch

vishnuchalla committed:
remove uneccessary log and modified load script to generate random layers for each run

vishnuchalla committed:
made changes in the load script to generate random layers

vishnuchalla committed:
adding clair setup custom yaml configuration

vishnuchalla committed:
filtering out non-emtpy items to avoid unecessary requests

vishnuchalla committed:
adding validity and mixed functionality for layers

vishnuchalla committed:
corrected mixed workload logic and assets file

crozzy commented on Makefile:
```diff
@@ -1,8 +1,52 @@
+# Containers
+ENGINE ?= podman
+REGISTRY = quay.io
+ORG ?= vchalla
```
I'd say don't include your personal namespace even as a default

vishnuchalla replied:
Acknowledged. I have used it just for my testing purposes, but happy to change it to a different repo which we are planning to use in the long term. Please feel free to suggest one. Thank you

crozzy commented on README.md:
```diff
@@ -1,15 +1,59 @@
+### **Usage on openshift platform**
+Deploy `assets/clair-config.yaml` on to your openshift cluster.
+### **Envs**
+* `CLAIR_TEST_HOST` - Sting indicating clair host to perform testing.
```
Nit `string`

vishnuchalla replied:
Changed as suggested. Thank you!

crozzy commented on README.md:
````diff
@@ -1,15 +1,59 @@
+```
+oc apply -f ~/assets/clair-config.yaml
+```
+Once deployed you should be able to see the details of the run in the pod logs with results getting logged and finally indexed to the target elsatic search index.
````
nit `elastic`

vishnuchalla replied:
Corrected as suggested!

crozzy commented on assets/clair-config.yaml:
```diff
@@ -0,0 +1,66 @@
+    spec:
+      containers:
+      - name: clair-testing
+        image: quay.io/vchalla/clair-load-test:amd64 
```
shouldn't be in your namespace going forward

vishnuchalla replied:
Acknowledged. I have used it just for my testing purposes, but happy to change it to a different repo which we are planning to use in the long term. Please feel free to suggest one. Thank you

crozzy replied:
`clair-load-test` you are now an admin

vishnuchalla replied:
Changed as suggested. Thank you!

crozzy commented on attacker/apis.go:
```diff
@@ -0,0 +1,98 @@
+
+// CreateIndexReportRequests returns the list of requests to perform POST operation on index_report.
+func CreateIndexReportRequests(ctx context.Context, manifests [][]byte, host, token string) []map[string]interface{} {
+	zlog.Debug(ctx).Msg(fmt.Sprintf("preparing %s requests for POST operation in index_report", strconv.Itoa(len(manifests))))
```
It's a little bit of an anti pattern to log like this, try and avoid formatting a string in the message and instead use the predefined methods with the correct types (it makes grepping the logs easier).

```go
zlog.Debug(ctx).Int("number of requests", len(manifests)).Msg("preparing requests for POST operation in index_report")
```

vishnuchalla replied:
Changed as suggested.

crozzy commented on manifests/manifests.go:
```diff
@@ -0,0 +1,81 @@
+			}
+		}
+	}()
+	wg.Wait()
```
This still does not make sense to me:
* What is the point in having a results channel and not just logging when you get an error, the only thing you're doing is iterating through a channel and logging errors.
* Why not close the `results` channel and then range over the results w/o another go routine for ranging.
* Why not use [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup)? (I don't remember if we had that discussion)

vishnuchalla replied:
Changed it to `errgroup` and removed the redundant channel as suggested.

crozzy commented on Makefile:
```diff
@@ -1,8 +1,52 @@
+	golangci-lint run
+
+install:
+	cp $(BIN_PATH) /usr/bin/$(BIN_NAME)
```
`install` also does this (and I believe has some permissions niceties)

vishnuchalla replied:
Yes. I have mentioned about this in `README.md`, simply adding a `sudo` should do it.

crozzy reviewed (changes requested):
Hey Vishnu, I left some more comments. 

Here are some observations from running it:
* The tests appear to experience 404 during initial runs, is the test waiting until the indexing is complete before carrying on to the subsequent requests?
* `Status Codes  [code:count]                      0:1  200:10` What is `0` status code?

vishnuchalla committed:
increasing timeout for vegeta tests

vishnuchalla committed:
api files

vishnuchalla committed:
merge testing into vchalla

vishnuchalla committed:
implemented vegeta using library and removed snafu dependency

vishnuchalla committed:
using errorgrp and resolving PR comments

vishnuchalla commented:
> Hey Vishnu, I left some more comments.
> 
> Here are some observations from running it:
> 
> * The tests appear to experience 404 during initial runs, is the test waiting until the indexing is complete before carrying on to the subsequent requests?
It was due to some outdated code. I have updated it now. It should be fixed by now and would request you to try re-running the tests. Thank you!
> * `Status Codes  [code:count]                      0:1  200:10` What is `0` status code?
In Vegeta, a status code of 0 typically indicates a network-related error or failure. It is not a standard HTTP status code. When using Vegeta as an HTTP load testing tool, a status code of 0 usually means that the request did not receive a valid HTTP response due to network issues, timeouts, or connection failures. Those are very less in number and can simply be ignored.

I have provided inline answers for your questions. Feel free to let me know if you have any questions further. Thank you!

vishnuchalla commented:
Hi @crozzy - Thanks for a quick review. Here is a summary of actions items completed.

- Vegeta Implementation using GO library - Completed.
- Replaced `UUID` as `RUNID` as it is not satisfying the `UUID` format.
- Replaced `waitGroup` with `errGroup` in fetching manifests.
- Fixed those nits and updated the `README.md`.

I only have one question, while trying to fetch 100 manifests, let's say we failed at 50th manifest, Do we want to gracefully proceed with load test logging the error for 50th one with those remaining 99 manifests or do we want to error out and stop the load test itself because of the failure we had?

vishnuchalla commented:
> Hi Crozzy - Thanks for a quick review. Here is a summary of actions items completed.
> 
> * Vegeta Implementation using GO library - Completed.
> * Replaced `UUID` as `RUNID` as it is not satisfying the `UUID` format.
> * Replaced `waitGroup` with `errGroup` in fetching manifests.
> * Fixed those nits and updated the `README.md`.
> 
> I only have one question, while trying to fetch 100 manifests, let's say we failed at 50th manifest, Do we want to gracefully proceed for the load test with those remaining 99 requests or do we want to error out and stop the load test itself?

crozzy commented:
good work, thanks for the perseverance, I think after updating the default quay.io org we should be good to go and can carry on iterating in `main`

vishnuchalla committed:
pointing to quay.io for load-test images

vishnuchalla commented:
> good work, thanks for the perseverance, I think after updating the default quay.io org we should be good to go and can carry on iterating in `main`

Updated the image to `quay.io` org as suggested.
